go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -38,77 +39,108 @@ import (
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
 )
 
+// We identify these heartbeat configuration types:
+// - No configuration: the heartbeat writer is generally disabled and does not produce heartbeats (but see below).
+// - On-demand: on-demand heartbeat interval was specified.
+// - Always: the heartbeat writer is always on and produces heartbeats at a regular interval.
+type HeartbeatConfigType int
+
+const (
+	HeartbeatConfigTypeNone HeartbeatConfigType = iota
+	HeartbeatConfigTypeOnDemand
+	HeartbeatConfigTypeAlways
+)
+
 const (
 	sqlUpsertHeartbeat = "INSERT INTO %s.heartbeat (ts, tabletUid, keyspaceShard) VALUES (%a, %a, %a) ON DUPLICATE KEY UPDATE ts=VALUES(ts), tabletUid=VALUES(tabletUid)"
+
+	testKeyspaceShard = "test:0"
+)
+
+var (
+	// Can be overridden by unit tests
+	defaultOnDemandDuration  = 10 * time.Second
```

Comment:
Why wouldn't we make this the default for the `--heartbeat_on_demand_duration` flag? That way the default is visible to users.

It also simplifies the logic here IMO, as then the two modes are `HeartbeatConfigTypeOnDemand` and `HeartbeatConfigTypeAlways`. It feels awkward to me that `HeartbeatConfigTypeNone` effectively becomes `HeartbeatConfigTypeOnDemand` and we use a default duration value that is not exposed/visible to users.

Reply:
> Why wouldn't we make this the default for the --heartbeat_on_demand_duration flag? 

Because if we made that the default, then that would change behavior to existing users, who perhaps do _not_ want to enable on-demand heartbeat, and want full heartbeat. 

That was also why we did not proceed with https://github.com/vitessio/vitess/pull/15099

Reply:
OK, then could we make/say that this flag is ignored when --enable_heartbeats is used? I think that's already the case isn't it? So I'm still not sure what we gain by doing things this way. Do you NOT have on demand heartbeats enabled by default as things are in the PR currently? I thought so. And if so, why would we not enforce the default duration via the flag instead of internally?

Reply:
> OK, then could we make/say that this flag is ignored when --heartbeat_enable is used?

Because again that would break existing setups. Think of the current configuration schema as an ancient mistake. Today we would have made the configuration different, but right now we need to stick to the existing rules.

> I think that's already the case isn't it? 

It is not.


Today, in `main`:

- If non of the flags is defined, there's no heartbeat whatsoever (which is what we're addressing here in this PR, enabling a lease of heartbeats by throttler's demand)
- If only `--heartbeat_enable` is set, then heartbeats are constant and always enabled.
- If `--heartbeat_on_demand_duration` is defined (whether `--heartbeat_enable` is set or not) then heartbeat is generally speaking enabled, but only producing heartbeats on-deman (and on open).

This means if someone only has `--heartbeat_enable` in their setup, and I set the flag default to `10s`, this would _reduce_ the generated heartbeats to on-demand duration only, even though the user expects them to be constant and continuous.

However, If we take your suggestion:

> could we make/say that this flag is ignored when --heartbeat_enable is used? 

Then people already using both flags `--heartbeat_enable` + `--heartbeat_on_demand_duration` and who are expecting only a limited lease of heartbeats, wil suddenly see an influx of heartbeats and a bloat of binary logs.

I hope this clarifies a bit.

Reply:
> Then people already using both flags --heartbeat_enable + --heartbeat_on_demand_duration and who are expecting only a limited lease of heartbeats, wil suddenly see an influx of heartbeats and a bloat of binary logs.

I thought that they both served the same purpose with regards to the throttler, the only thing `--heartbeat_enable` got you was sub-second replication lag metrics via the replication reporter vs. instead using the polling based reporter via `--enable_replication_reporter`.  Perhaps I'm wrong on that part?

I thought that using both flags would also potentially lead to failures? https://github.com/vitessio/vitess/issues/14978

I was thinking that 1) (virtually) nobody is using both 2) if they are, they shouldn't be as that can case failures 3) we could take this opportunity to clean this up a bit. Maybe I misunderstood though? 

In any event, you've explained that it's more complicated than I had initially thought and we don't need to solve these issues in this PR. 🙂 Thanks!

Reply:
> we could take this opportunity to clean this up a bit. 

I'm in favor. I don't like the existing configuration setup. If we can change things in one shot and without disrupting existing setups, I'm happy to. In which case, as you say, the changes in this PR can be greatly simplified. Let's perhaps make that discussion tomorrow.